### PR TITLE
tools/cephfs: add accounted_rstat/rstat when building file dentry

### PR DESCRIPTION
**Pull request by lxbsz**
When the file metadata is corrupted and doing the file recovery,
let trust the dirfrag's unmatched rstat info.

https://tracker.ceph.com/issues/44380
Signed-off-by: Xiubo Li <xiubli@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by ukernel** [comment]
I don't think this change is correct.  parent node's rstat is accumulation of child nodes' accounted_rstat. Locker::check_inode_max_size() is not special, like file size update and truncate, it update inode's rstat.

**Comment by lxbsz** [comment]
> I don't think this change is correct. parent node's rstat is accumulation of child nodes' accounted_rstat. Locker::check_inode_max_size() is not special, like file size update and truncate, it update inode's rstat.

@ukernel 
Updated it, the root cause it the data scan tool didn't update the accounted_rstat/rstat.

**Comment by ukernel** [comment]
I don't think it's worth to make data scan tool update rstat. because there is no guarantee that original rstats on inodes/dirfrags are correct

**Comment by lxbsz** [comment]
> I don't think it's worth to make data scan tool update rstat. because there is no guarantee that original rstats on inodes/dirfrags are correct. after data scan, we should run a scrub

The scrub will adjust the "subdir/" inode's rstat from dirfrag, and the rstat is not correct, it is b188, but b190 is expected:

```
# tree -h /tmp/mydir/
/tmp/mydir/
└── [ 188]  subdir
    ├── [   1]  0
    ├── [   1]  1
    ├── [   2]  10
    ├── [   2]  11
    ├── [   2]  12
    ├── [   2]  13
    ├── [   2]  14
    ├── [   2]  15
    ├── [   2]  16
    ├── [   2]  17
    ├── [   2]  18
    ├── [   2]  19
    ├── [   1]  2
    ├── [   2]  20
    ├── [   2]  21
    ├── [   2]  22
    ├── [   2]  23
    ├── [   2]  24
    ├── [   2]  25
    ├── [   2]  26
    ├── [   2]  27
    ├── [   2]  28
    ├── [   2]  29
    ├── [   1]  3
    ├── [   2]  30
    ├── [   2]  31
    ├── [   2]  32
    ├── [   2]  33
    ├── [   2]  34
    ├── [   2]  35
    ├── [   2]  36
    ├── [   2]  37
    ├── [   2]  38
    ├── [   2]  39
    ├── [   1]  4
    ├── [   2]  40
    ├── [   2]  41
    ├── [   2]  42
    ├── [   2]  43
    ├── [   2]  44
    ├── [   2]  45
    ├── [   2]  46
    ├── [   2]  47
    ├── [   2]  48
    ├── [   2]  49
    ├── [   1]  5
    ├── [   2]  50
    ├── [   2]  51
    ├── [   2]  52
    ├── [   2]  53
    ├── [   2]  54
    ├── [   2]  55
    ├── [   2]  56
    ├── [   2]  57
    ├── [   2]  58
    ├── [   2]  59
    ├── [   1]  6
    ├── [   2]  60
    ├── [   2]  61
    ├── [   2]  62
    ├── [   2]  63
    ├── [   2]  64
    ├── [   2]  65
    ├── [   2]  66
    ├── [   2]  67
    ├── [   2]  68
    ├── [   2]  69
    ├── [   1]  7
    ├── [   2]  70
    ├── [   2]  71
    ├── [   2]  72
    ├── [   2]  73
    ├── [   2]  74
    ├── [   2]  75
    ├── [   2]  76
    ├── [   2]  77
    ├── [   2]  78
    ├── [   2]  79
    ├── [   1]  8
    ├── [   2]  80
    ├── [   2]  81
    ├── [   2]  82
    ├── [   2]  83
    ├── [   2]  84
    ├── [   2]  85
    ├── [   2]  86
    ├── [   2]  87
    ├── [   2]  88
    ├── [   2]  89
    ├── [   1]  9
    ├── [   2]  90
    ├── [   2]  91
    ├── [   2]  92
    ├── [   2]  93
    ├── [   2]  94
    ├── [   2]  95
    ├── [   2]  96
    ├── [   2]  97
    ├── [   2]  98
    └── [   2]  99

1 directory, 100 files
```

**Comment by lxbsz** [comment]
> > I don't think it's worth to make data scan tool update rstat. because there is no guarantee that original rstats on inodes/dirfrags are correct. after data scan, we should run a scrub
> 
> The scrub will adjust the "subdir/" inode's rstat from dirfrag, and the rstat is not correct, it is b188, but b190 is expected:
> 
> ```
> # tree -h /tmp/mydir/
> /tmp/mydir/
> └── [ 188]  subdir
>     ├── [   1]  0
>     ├── [   1]  1
...
> 1 directory, 100 files
> ```

But in this test case IMO it should be okay, I will update it with running a scrub here.

**Comment by lxbsz** [comment]
@ukernel 
Updated it, thanks.

**Comment by ukernel** [comment]
> But in this test case IMO it should be okay, I will update it with running a scrub here.

I think disable auto dirfrag split/merge fix for this case is better.

File: src/tools/cephfs/DataScan.h [hunk 1]
```diff
@@ -220,7 +220,7 @@ class MetadataDriver : public RecoveryDriver, public MetadataTool
 
     int inject_linkage(
         inodeno_t dir_ino, const std::string &dname,
-        const frag_t fragment, const InodeStore &inode, snapid_t dnfirst=CEPH_NOSNAP);
+        const frag_t fragment, const InodeStore &inode, snapid_t dnfirst=2);
```
Comment (ukernel) [inline src/tools/cephfs/DataScan.h:223]: 'cephfs-data-scan scan_links' should fix this.

Comment (lxbsz) [reply src/tools/cephfs/DataScan.h:223]: Tried, it is still the 'head' and it crashes.

Comment (lxbsz) [reply src/tools/cephfs/DataScan.h:223]: Sorry, it works, it is 2.

**Comment by lxbsz** [comment]
> > But in this test case IMO it should be okay, I will update it with running a scrub here.
> 
> I think disable auto dirfrag split/merge fix for this case is better.

BTW, what does the auto dirfrag spli/merge mean here ?

I think this test case was allowed the rstat info is different between the dirfrag and inode, this should be why it disables :

```
self.fs.set_ceph_conf('mds', 'mds verify scatter', Flase)                                          
self.fs.set_ceph_conf('mds', 'mds debug scatterstat', False) 
```

But after this test case, the subdir/ directory was not removed and in other the followed cases the above 2 config are enabled and at the same time the "subdir/21" was set to the recovery queue. This should be why the MDS daemon crashing hit in Patrick's test.

**Comment by ukernel** [comment]
> > > But in this test case IMO it should be okay, I will update it with running a scrub here.
> > 
> > 
> > I think disable auto dirfrag split/merge fix for this case is better.
> 
> BTW, what does the auto dirfrag spli/merge mean here ?

set config 'mds_bal_merge_size' to 0

**Comment by lxbsz** [comment]
> > > > But in this test case IMO it should be okay, I will update it with running a scrub here.
> > > 
> > > 
> > > I think disable auto dirfrag split/merge fix for this case is better.
> > 
> > 
> > BTW, what does the auto dirfrag spli/merge mean here ?
> 
> set config 'mds_bal_merge_size' to 0
> set config 'mds_bal_merge_size' to a very large valye

Okay, will try it.
Thanks

**Comment by lxbsz** [comment]
@ukernel 
This will add:
1, scan links   ===> will fix the ceph_assert(in->first <= straydn->first)
2, scrub the dir        ==> this will avoid the verify scatter crash issue
3, disable the auto merge/split explicitly  ==> avoid the auto merge/split noise
4, remove the /subdir

**Comment by gregsfortytwo** [comment]
http://pulpito.front.sepia.ceph.com/gregf-2020-04-15_20:54:50-fs-wip-greg-testing-415-1-distro-basic-smithi/ failures are https://tracker.ceph.com/issues/45133, https://tracker.ceph.com/issues/45103, and a couple rpm metadata download errors.

**Pull request merged**